src/Pos/Launcher/Resource.hs

```diff
@@ -160,27 +194,49 @@ allocateNodeResources np@NodeParams {..} sscnp = do
             , ..
             }
 
+-- | An 'm' terms contains enough information to discharge any other 'm' term
+--   to some base 'b'. Canonical example is the instance
+--
+--     'LowerM (ReaderT r b) b'
+--
+--   which uses its own reader antecedent to discharge other 'ReaderT r'
+--   constructors.
+class LowerM m b where
+    lowerM :: m (m t -> b t)
+
+instance Monad b => LowerM (ReaderT r b) b where
+    lowerM = do
+        r <- ask
+        return $ flip runReaderT r
+
 -- | Release all resources used by node. They must be released eventually.
 releaseNodeResources ::
        forall ssc m. (SscConstraint ssc, MonadIO m)
-    => NodeResources ssc m -> m ()
+    => NodeResources ssc m -> Production ()
 releaseNodeResources NodeResources {..} = do
     releaseAllHandlers
     whenJust nrJLogHandle (liftIO . hClose)
     closeNodeDBs nrDBs
     releaseNodeContext nrContext
-    closeTransport nrTransport
 
 -- | Run computation which requires 'NodeResources' ensuring that
 -- resources will be released eventually.
-bracketNodeResources :: forall ssc a.
-      (SscConstraint ssc, SecurityWorkersClass ssc)
+bracketNodeResources :: forall ssc m a.
+      ( SscConstraint ssc
+      , SecurityWorkersClass ssc
+      , WithLogger m
+      , MonadIO m
+      , LowerM m Production
+      , MonadMockable m
+      )
     => NodeParams
     -> SscParams ssc
-    -> (NodeResources ssc Production -> Production a)
+    -> (NodeResources ssc m -> Production a)
     -> Production a
-bracketNodeResources np sp =
-    bracket (allocateNodeResources np sp) releaseNodeResources
+bracketNodeResources np sp k = bracketTransport tcpAddr $ \transport ->
```

Comment:
The monad in which we allocate the transport is not necessarily the same as the monad in which we use it. For instance, we can allocate a typical TCP transport in `IO`, then lift it via `Network.Transport.Concrete.concrete` to any `MonadIO`. By moving it outside, we can de-allocate it (`NT.close) in the same monad in which it was created.